### PR TITLE
initial vsi module added

### PR DESCRIPTION
**Pull request by VaishnaviGopal**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,26 @@
+###################################################################
+# Instance module usage
+###################################################################
+
+module "vsi" {
+  //Uncomment the following line to point the source to registry level
+  //source  = "terraform-ibm-modules/observability/ibm//modules/instance"
```
Comment (umarali-nagoor) [inline main.tf]: Change the source

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  source = "./modules/vsi"
+
+  count = var.instance_count
+  provision            = var.vsi_provision
```
Comment (umarali-nagoor) [inline main.tf]: Prefix variable names with "is_" for boolean variables

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  name                 = format("%s-logging-str", var.logging_name)
+  bind_key             = var.str_logging_bind_key
+  enable_platform_logs = var.enable_platform_logs
+  resource_group_id    = var.resource_group_id
+  plan                 = var.str_logging_plan
+  region               = var.str_logging_region
```
Comment (umarali-nagoor) [inline main.tf]: Sorry, Why are we having STR related parameters ?

File: modules/floatingIP/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+## Example Usage
+```
+
+data "ibm_resource_group" "resource_group" {
+  name = (var.resource_group != null ? var.resource_group : "default")
+}
+
+module "fip" {
+  source = "terraform-ibm-modules/vpc/ibm//modules/floatingIP"
```
Comment (umarali-nagoor) [inline modules/floatingIP/README.md]: Why are we referring to VPC repo ? It is going to be deprecated.

File: modules/floatingIP/variables.tf [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#####################################################
+# Floating IP Parameters
+# Copyright 2020 IBM
+#####################################################
+
+variable "provision" {
```
Comment (umarali-nagoor) [inline modules/floatingIP/variables.tf]: Suffix boolean varaibles with "is_"

File: modules/ssh-key/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+This module is used to create an SSH Key
+
+## Example Usage
+```
+data "ibm_resource_group" "resource_group" {
+  name = (var.resource_group != null ? var.resource_group : "default")
+}
+
+module "ssh-key" {
+  source = "terraform-ibm-modules/vpc/ibm//modules/ssh-key"
```
Comment (umarali-nagoor) [inline modules/ssh-key/README.md]: change the source path

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  encryption        = each.value.encryption
+  tags              = each.value.tags
+}
+
+locals {
+  sshkeys = [for key in module.sshkey : key.ssh_key_id]
+  volumes = [for volume in module.storage : volume.volume_id]
+}
+
+module "instance" {
```
Comment (umarali-nagoor) [inline main.tf:4]: Please check missing arguments  like
default_trusted_profile_auto_link
crn
port_speed

make sure we cover all mandatory and optional arguments

File: modules/vsi/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+| tags | list of tags for the instance  | list(string) | n/a | no |
+| wait_before_delete | Enables stopping of instance before deleting and waits till deletion is complete | bool | n/a | no |
+| action | Action to be taken on the instance. Supported values are stop, start, or reboot | string | n/a | no |
+| force_action | If set to true, the action will be forced immediately, and all queued actions deleted. Ignored for the start action | string | n/a | no|
+| primary\_network\_interface | List of primary_network_interface that are to be attached to the instance  | list(object) | n/a | no |
+| network\_interfaces | List of network_interfaces that are to be attached to the instance  | list(object) | n/a | no |
+| user\_data | User Data for the instance  | string | n/a | no |
+| image | Image ID for the instance  | string | n/a | no |
+| boot\_volume | List of boot volume that are to be attached to the instance| list(object) | n/a | no |
+| volumes | List of volume ids that are to be attached to the instance  | list(string) | n/a | no |
```
Comment (umarali-nagoor) [inline modules/vsi/README.md:62]: Describe the list(object) variables information in a separate table like primary_network_interface etc

File: README.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# IBM Cloud <module_name> - Terraform Module
+# IBM Cloud VPC VSI - Terraform Module
```
Comment (umarali-nagoor) [inline README.md:1]: Update test.yml file with code to write test results into slack channel (Take code from https://github.com/terraform-ibm-modules/terraform-ibm-module-template repo)

File: README.md [hunk 10]
```diff
@@ -79,3 +77,45 @@ terraform destroy -var-file=./input.tfvars
 
 All optional parameters, by default, will be set to `null` in respective example's variable.tf file. You can also override these optional parameters.
 
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Inputs
+
```
Comment (umarali-nagoor) [inline README.md:82]: Add following hook in validate.yml github action (Take code from https://github.com/terraform-ibm-modules/terraform-ibm-module-template repo)

auto-update hook

File: test/vsi_test.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+		// The path to where our Terraform code is located
+		TerraformDir: "../examples/vsi",
+
+		Vars: map[string]interface{}{
+			"name":                "test-vsi",
+			"vpc_name":            "test-vsi-vpc",
+			"subnet_name":         "test-vsi-subnet",
+			"location":            "us-south-2",
+			"image":               "r006-5b05b4fe-bcbc-4309-ad45-3354813227a0",
+			"profile":             "cx2-2x4",
```
Comment (umarali-nagoor) [inline test/vsi_test.go:24]: in vsi_test.go file you have hardcoded values like
"image":               "r006-5b05b4fe-bcbc-4309-ad45-3354813227a0",
"ssh_keys":            "ssh-rsa AAAAB3NzaC1yc2EAAAADAQA"
May work locally as you have these image and ssh_key locally in your machine
but test case fails once you push changes to master branch as these values will not present in repo where we run this test case

File: examples/vsi-with-storage/main.tf [hunk 12]
```diff
@@ -0,0 +1,32 @@
+#####################################################
+# vsi with storage Example
+# Copyright 2022 IBM
+#####################################################
+
+module "volume" {
```
Comment (umarali-nagoor) [inline examples/vsi-with-storage/main.tf:6]: Add detect secret hook in pre-commit file (Take code from https://github.com/terraform-ibm-modules/terraform-ibm-module-template repo)
https://github.com/VaishnaviGopal/terraform-ibm-vpc-vsi/blob/main/.pre-commit-config.yaml

File: examples/vsi-with-storage/variables.tf [hunk 13]
```diff
@@ -0,0 +1,85 @@
+#####################################################
+# vsi with storage parameters
+# Copyright 2022 IBM
+#####################################################
+
+variable "volume_list" {
```
Comment (umarali-nagoor) [inline examples/vsi-with-storage/variables.tf:6]: Update changelog with your current changes
https://github.com/VaishnaviGopal/terraform-ibm-vpc-vsi/blob/main/CHANGELOG.md

File: test/vsi_test.go (hunk 11 above)
Comment (VaishnaviGopal) [reply test/vsi_test.go:24]: Irrespective of account this works. so no issue with these

File: examples/vsi-with-storage/variables.tf (hunk 13 above)
Comment (VaishnaviGopal) [reply examples/vsi-with-storage/variables.tf:6]: done.

File: examples/vsi-with-storage/main.tf (hunk 12 above)
Comment (VaishnaviGopal) [reply examples/vsi-with-storage/main.tf:6]: done

File: README.md (hunk 10 above)
Comment (VaishnaviGopal) [reply README.md:82]: done

File: README.md (hunk 9 above)
Comment (VaishnaviGopal) [reply README.md:1]: done

File: main.tf (hunk 7 above)
Comment (VaishnaviGopal) [reply main.tf:4]: yes. have covered all of the parameters. These might have been added very recently after the developement phase. will be addressing them in the upcoming PR

File: modules/vsi/README.md (hunk 8 above)
Comment (VaishnaviGopal) [reply modules/vsi/README.md:62]: Sure. will address in the upcoming PR after the initial release

**Pull request merged**